Vyatta specific preemption code

Code to pass Vyatta specific regressions

---
 keepalived/vrrp/vrrp.c |   16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

--- a/keepalived/vrrp/vrrp.c
+++ b/keepalived/vrrp/vrrp.c
@@ -4868,7 +4868,15 @@ restore_vrrp_state(vrrp_t *old_vrrp, vrr
 	bool added_ip_addr = false;
 
 	/* If the new state is master, we must be reloading from master */
-	vrrp->reload_master = vrrp->state == VRRP_STATE_MAST;
+	if(vrrp->state == VRRP_STATE_MAST || (old_vrrp->state == VRRP_STATE_MAST && vrrp->nopreempt && vrrp->base_priority != old_vrrp->base_priority)) {
+		vrrp->reload_master = true;
+		vrrp->state =
+#ifdef _WITH_SNMP_VRRP
+		vrrp->configured_state =
+#endif
+		vrrp->wantstate = VRRP_STATE_MAST;
+	} else
+		vrrp->reload_master = false;
 
 	/* Save old stats */
 	memcpy(vrrp->stats, old_vrrp->stats, sizeof(vrrp_stats));
@@ -4904,6 +4912,12 @@ restore_vrrp_state(vrrp_t *old_vrrp, vrr
 #endif
 	}
 
+	if (!vrrp->reload_master)
+		vrrp->state =
+#ifdef _WITH_SNMP_VRRP
+		vrrp->configured_state =
+#endif
+		vrrp->wantstate = VRRP_STATE_BACK;
 	return added_ip_addr;
 }
 
